### PR TITLE
Feat/addr

### LINKED ISSUE #75: addr

### PR DESCRIPTION
**Pull request by oxlime**
should resolve #75

### CODE REVIEW

File: src/network/protocol/messages/addr.zig [hunk 1]
```diff
@@ -0,0 +1,180 @@
+const std = @import("std");
+const native_endian = @import("builtin").target.cpu.arch.endian();
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:2]: not used

File: src/network/protocol/messages/addr.zig [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+const protocol = @import("../lib.zig");
+
+const ServiceFlags = protocol.ServiceFlags;
+
+const Endian = std.builtin.Endian;
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:7]: not used

File: src/network/protocol/messages/addr.zig [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+const Sha256 = std.crypto.hash.sha2.Sha256;
+
+const CompactSizeUint = @import("bitcoin-primitives").types.CompatSizeUint;
+
+/// AddrMessage represents the "addr" message
+///
+/// https://developer.bitcoin.org/reference/p2p_networking.html#addr
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:14]: this doc goes on the AddrMessage struct

File: src/network/protocol/messages/addr.zig [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            if (!std.meta.hasFn(@TypeOf(w), "writeAll")) @compileError("Expects r to have fn 'writeAll'.");
+        }
+        //try CompactSizeUint.new(@intCast(self.ip_addresses.len)).encodeToWriter(w);
+        try CompactSizeUint.new(self.ip_addresses.len).encodeToWriter(w);
+
+        // Serialize each IP address
+        for (self.ip_addresses) |*ip_address| {
+            try w.writeInt(u32, ip_address.time, .little);
+            try w.writeInt(u64, ip_address.services, .little);
+        try w.writeAll(std.mem.asBytes(&ip_address.ip));
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:73]: indentation

File: src/network/protocol/messages/addr.zig [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !AddrMessage {
+        var fbs = std.io.fixedBufferStream(bytes);
+        const reader = fbs.reader();
+        return try AddrMessage.deserializeReader(allocator, reader);
+    }
+
+    pub fn hintSerializedLen(self: AddrMessage) usize {
+        // 4 + 8 + 16 + 2
+        const fixed_length_per_ip = 30;
+        return 1 +  self.ip_addresses.len * fixed_length_per_ip;// 1 for CompactSizeUint
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:135]: CompactSizeUint len is not 1, it's CompactSizeUint.new(ip_addresses.len).hint_encoded_size()

The bigger the number, the more bytes it takes to write it

File: src/network/protocol/messages/addr.zig [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    pub fn eql(self: *const AddrMessage, other: *const AddrMessage) bool {
+        if (self.ip_addresses.len != other.ip_addresses.len) return false;
+
+    const count = @as(usize, self.ip_addresses.len); 
+    for (0..count) |i| {
+        if (!self.ip_addresses[i].eql(&other.ip_addresses[i])) return false;
+    }
+
+    return true;
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:147]: indentation

Comment (tdelabro) [reply src/network/protocol/messages/addr.zig:147]: run `zig fmt`

File: src/network/wire/lib.zig [hunk 7]
```diff
@@ -19,6 +19,9 @@ pub const Error = error{
     MessageTooLarge,
 };
 
+pub const NetworkIPAddr = @import("../protocol/messages/addr.zig").NetworkIPAddr;
+const CompactSizeUint = @import("bitcoin-primitives").types.CompatSizeUint;
+
```
Comment (tdelabro) [inline src/network/wire/lib.zig:24]: not used, or only inside your test, if so move them there

File: src/network/wire/lib.zig [hunk 8]
```diff
@@ -329,6 +334,62 @@ test "ok_send_pong_message" {
     }
 }
 
+test "ok_send_addr_message" {
+    const Config = @import("../../config/config.zig").Config;
+
+    const ArrayList = std.ArrayList;
+    const test_allocator = std.testing.allocator;
+    const AddrMessage = protocol.messages.AddrMessage;
+    //const ServiceFlags = protocol.ServiceFlags;
```
Comment (tdelabro) [inline src/network/wire/lib.zig:343]: remove

File: src/network/wire/lib.zig [hunk 9]
```diff
@@ -329,6 +334,62 @@ test "ok_send_pong_message" {
...
+//    const received_message = try receiveMessage(test_allocator, reader);
+//    defer received_message.deinit(test_allocator);
+//
+//    switch (received_message) {
+//        .Addr => |rm| try std.testing.expect(message.eql(&rm)),
+//        .Version => unreachable,
+//        .Verack => unreachable,
+//        .Mempool => unreachable,
+//        .Getaddr => unreachable,
+//    }
```
Comment (tdelabro) [inline src/network/wire/lib.zig:390]: forgot to remove this

**Review by tdelabro — CHANGES_REQUESTED** [review]
Good job!

File: src/network/protocol/messages/addr.zig [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        const ip_addresses = try allocator.alloc(NetworkIPAddr, ip_address_count.value());
+
+        vm.ip_addresses = ip_addresses;
+
+        for (vm.ip_addresses) |*ip_address| {
+            ip_address.time = try r.readInt(u32, .little);
+            ip_address.services = try r.readInt(u64, .little);
+            try r.readNoEof(&ip_address.ip);
+            ip_address.port = try r.readInt(u16, .big);
+        }
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:113]: create a NetworkIpAddress.deserializeReader method

File: src/network/protocol/messages/addr.zig [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        //try CompactSizeUint.new(@intCast(self.ip_addresses.len)).encodeToWriter(w);
+        try CompactSizeUint.new(self.ip_addresses.len).encodeToWriter(w);
+
+        // Serialize each IP address
+        for (self.ip_addresses) |*ip_address| {
+            try w.writeInt(u32, ip_address.time, .little);
+            try w.writeInt(u64, ip_address.services, .little);
+            try w.writeAll(std.mem.asBytes(&ip_address.ip));
+            try w.writeInt(u16, ip_address.port, .big);
+        }
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:68]: same

File: src/network/protocol/messages/addr.zig [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+    /// Serialize the message as bytes and write them to the Writer.
+    ///
+    /// `w` should be a valid `Writer`.
+    pub fn serializeToWriter(self: *const AddrMessage, w: anytype) !void {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(w), "writeInt")) @compileError("Expects r to have fn 'writeInt'.");
+            if (!std.meta.hasFn(@TypeOf(w), "writeAll")) @compileError("Expects r to have fn 'writeAll'.");
+        }
+        //try CompactSizeUint.new(@intCast(self.ip_addresses.len)).encodeToWriter(w);
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:59]: forgotten comment

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by tdelabro** [comment]
@oxlime are you still working on this?

File: src/network/protocol/messages/addr.zig [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    services: u64, // Services offered by the node
+    ip: [16]u8, // IPv6 address (including IPv4-mapped)
+    port: u16, // Port number
+
+    // NetworkIPAddr eql
+    pub fn eql(self: *const NetworkIPAddr, other: *const NetworkIPAddr) bool {
+        return self.time == other.time and self.services == other.services and std.mem.eql(u8, &self.ip, &other.ip) and self.port == other.port;
+    }
+
+    pub fn serializeWriter(self: *const NetworkIPAddr, writer: anytype) !void {
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:20]: serializeToWriter

File: src/network/protocol/messages/addr.zig [hunk 14]
```diff
@@ -0,0 +1,180 @@
+const std = @import("std");
+const protocol = @import("../lib.zig");
+
+const Endian = std.builtin.Endian;
+const Sha256 = std.crypto.hash.sha2.Sha256;
+
+const CompactSizeUint = @import("bitcoin-primitives").types.CompatSizeUint;
+
+pub const NetworkIPAddr = struct {
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:9]: part of it already exists under src/network/protocol/NetworkAddress.zig

please use it

Comment (tdelabro) [reply src/network/protocol/messages/addr.zig:9]: You can reorder the fields if you want

Comment (tdelabro) [reply src/network/protocol/messages/addr.zig:9]: so you have a struct which is time + NetworkAddress

File: src/network/protocol/messages/addr.zig [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+    ///
+    /// Computed as `Sha256(Sha256(self.serialize()))[0..4]`
+    pub fn checksum(self: AddrMessage) [4]u8 {
+        var digest: [32]u8 = undefined;
+        var hasher = Sha256.init(.{});
+        const writer = hasher.writer();
+        self.serializeToWriter(writer) catch unreachable; // Sha256.write is infaible
+        hasher.final(&digest);
+
+        Sha256.hash(&digest, &digest, .{});
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:57]: use generic checksum

File: src/network/protocol/messages/addr.zig [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        allocator.free(self.ip_addresses);
+    }
+
+    /// Serialize the message as bytes and write them to the Writer.
+    ///
+    /// `w` should be a valid `Writer`.
+    pub fn serializeToWriter(self: *const AddrMessage, w: anytype) !void {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(w), "writeInt")) @compileError("Expects r to have fn 'writeInt'.");
+            if (!std.meta.hasFn(@TypeOf(w), "writeAll")) @compileError("Expects r to have fn 'writeAll'.");
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:73]: you are using none of those. so don't check

File: src/network/protocol/messages/addr.zig [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+
+        return ret;
+    }
+
+    /// Deserialize a Reader bytes as a `AddrMessage`
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !AddrMessage {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects r to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects r to have fn 'readNoEof'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readAll")) @compileError("Expects r to have fn 'readAll'.");
```
Comment (tdelabro) [inline src/network/protocol/messages/addr.zig:107]: same

File: src/network/protocol/messages/lib.zig [hunk 18]
```diff
@@ -145,6 +150,7 @@ pub const Message = union(MessageTypes) {
             .notfound => |*m| m.checksum(),
             .sendheaders => |*m| m.checksum(),
             .filterload => |*m| m.checksum(),
+            .addr => |m| m.checksum(),
```
Comment (tdelabro) [inline src/network/protocol/messages/lib.zig:153]: |*m|

File: src/network/protocol/messages/lib.zig [hunk 19]
```diff
@@ -166,6 +172,7 @@ pub const Message = union(MessageTypes) {
             .notfound => |m| m.hintSerializedLen(),
             .sendheaders => |m| m.hintSerializedLen(),
             .filterload => |*m| m.hintSerializedLen(),
+            .addr => |m| m.hintSerializedLen(),
```
Comment (tdelabro) [inline src/network/protocol/messages/lib.zig:175]: |*m|

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Review by tdelabro — APPROVED** [review]

**Comment by tdelabro** [comment]
@oxlime just rebase and I will merge

**Comment by tdelabro** [comment]
@oxlime sorry you need to fix conflicts again :,(

File: src/network/protocol/messages/addr.zig [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    ///
+    /// buffer.len must be >= than self.hintSerializedLen()
+    pub fn serializeToSlice(self: *const AddrMessage, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        const writer = fbs.writer();
+        try self.serializeToWriter(writer);
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const AddrMessage, allocator: std.mem.Allocator) ![]u8 {
```
Comment (ybensacq) [inline src/network/protocol/messages/addr.zig:75]: you can use the generic function

File: src/network/protocol/messages/addr.zig [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+            ip_address.* = try NetworkIPAddr.deserializeReader(r);
+        }
+
+        return AddrMessage{
+            .ip_addresses = ip_addresses,
+        };
+    }
+
+    /// Deserialize bytes into a `AddrMessage`
+    pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !AddrMessage {
```
Comment (ybensacq) [inline src/network/protocol/messages/addr.zig:104]: we have a generic function for this one as well.

**Review by tdelabro — APPROVED** [review]

**Pull request merged**